main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -893,6 +893,19 @@ object Typer {
           resultEff = Type.mkUnion(eff :: guardEffs ::: bodyEffs, loc)
         } yield (constrs ++ guardConstrs.flatten ++ bodyConstrs.flatten, resultTyp, resultPur, resultEff)
 
+      case KindedAst.Expression.MatchType(exp, rules, loc) =>
+        val bodies = rules.map(_.exp)
+
+        for {
+          (constrs, tpe, pur, eff) <- visitExp(exp)
+          // unify each rule's variable with its type
+          _ <- seqM(rules.map(rule => unifyTypeM(rule.sym.tvar, rule.tpe, rule.sym.loc)))
```

Comment:
I think we have to be careful with rigidity. Imagine if one writes:

```
typematch x {
  case v: List[a] => v == 123 :: Nil
}
```

Reply:
`a` is necessarily in the `renv` here, so this will not compile. I'll add a test.